src/federationparams.h

```diff
@@ -35,14 +40,42 @@ class CFederationParams
     /**
      * Parse aggPubkey in block header.
      */
-    CPubKey ReadAggregatePubkey(const std::vector<unsigned char>& pubkey);
-    const CPubKey& GetAggregatePubkey() const{ return aggregatePubkey; }
+    CPubKey ReadAggregatePubkey(const std::vector<unsigned char>& pubkey, uint height);
+    const CPubKey& GetLatestAggregatePubkey() const { return aggregatePubkeyHeight.back().aggpubkey; }
     bool ReadGenesisBlock(std::string genesisHex);
     const CBlock& GenesisBlock() const { return genesis; }
     const std::string& getDataDir() const { return dataDir; }
     const std::vector<SeedSpec6>& FixedSeeds() const { return vFixedSeeds; }
     /** Return the list of hostnames to look up for DNS seeds */
     const std::vector<std::string>& DNSSeeds() const { return vSeeds; }
+    const uint& GetHeightFromAggregatePubkey(std::vector<unsigned char> aggpubkey) const {
+        for (auto& c : aggregatePubkeyHeight)
+            if (c.aggpubkey == CPubKey(aggpubkey.begin(), aggpubkey.end())) {
+                return c.height;
+                break;
+            } else {
+                continue;
+            }
+    };
+    const CPubKey& GetAggPubkeyFromHeight(uint height) const {
```

Comment:
`if((height < 1) || (aggregatePubkeyHeight.size() == 1))` double parentheses '((' are not necessary.
`decltype(aggregatePubkeyHeight.size())` is accurate, but unnecessary, we can use `unsigned int `instead. 

I'm not sure if it is safe to access element (i+1) in a loop running from i = 0 to size - 1.  for new blocks, when i is size - 1, we'll try to compare it to the height at 'size' element which is out of the vector. we can eliminate many lines from the loop and handle that case with this reordering:
```
const CPubKey& GetAggPubkeyFromHeight(uint height) const 
{
    if(height < 1 || aggregatePubkeyHeight.size() == 1) {
        return aggregatePubkeyHeight.at(0).aggpubkey; 

    if(height > aggregatePubkeyHeight.end().height)
        return aggregatePubkeyHeight.end().aggpubkey;

    for(int i = 0; i < aggregatePubkeyHeight.size(); i++) {
        if(height == aggregatePubkeyHeight.at(i).height)
            return aggregatePubkeyHeight.at(i).aggpubkey;

        else if(aggregatePubkeyHeight.at(i).height < height) && (height < aggregatePubkeyHeight.at(i+1).height)) 
            return aggregatePubkeyHeight.at(i).aggpubkey;
    }
}
```

Reply:
in your for loop look like both the if and else if return the same thing 
so i group them together